HCP docs update for IBM Power (#92359)
openshift/openshift-docs: modules/hcp-ibm-power-create-heterogeneous-nodepools-agent-hc.adoc

Reviewer: To comply with the modular documentation guidelines that we follow for the OCP documentation, the various sections in this file need to be moved to their own modules. After you create those modules, in the assembly file, add the modules below your new `include::modules/hcp-ibm-power-create-heterogeneous-nodepools-agent-hc.adoc[leveloffset=+1]` module, and ensure that instead of using `[leveloffset=+1]`, you use `[leveloffset=+2]`.

// Module included in the following assemblies:
//
// * hosted_control_planes/hcp-deploy/hcp-deploy-ibm-power.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-ibm-power-create-heterogeneous-nodepools-agent-hc_{context}"]
= Create Heterogeneous NodePools on Agent HostedClusters

This explains how to create heterogeneous nodepools on an agent platform. Please refer to set up the environment using Agent Platform from section 4.6.1, this section only covers the things needed to configure to have heterogeneous nodepools.

.Procedure

. Configure AgentServiceConfig with two heterogeneous architecture OS images:

> +

Reviewer: From the [OCP docs guidelines](https://github.com/openshift/openshift-docs/pull/92359): Do not use more than one command per code block.

$ export OCP_RELEASE_VERSION_X86=$(curl -s https://mirror.openshift.com/pub/openshift-v4/${ARCH_X86}/clients/ocp/${OCP_VERSION}/release.txt | awk '/machine-os / { print $2 }')
$ export ISO_URL_X86="https://mirror.openshift.com/pub/openshift-v4/${ARCH_X86}/dependencies/rhcos/${OCP_MAJMIN}/${OCP_VERSION}/rhcos-${OCP_VERSION}-${ARCH_X86}-live.${ARCH_X86}.iso"
$ export ROOT_FS_URL_X8="https://mirror.openshift.com/pub/openshift-v4/${ARCH_X86}/dependencies/rhcos/${OCP_MAJMIN}/${OCP_VERSION}/rhcos-${OCP_VERSION}-${ARCH_X86}-live-rootfs.${ARCH_X86}.img"
$ export ARCH_PPC64LE="ppc64le"
$ export OCP_RELEASE_VERSION_PPC64LE=$(curl -s https://mirror.openshift.com/pub/openshift-v4/${ARCH_PPC64LE}/clients/ocp/${OCP_VERSION}/release.txt | awk '/machine-os / { print $2 }')
$ export ISO_URL_PPC64LE="https://mirror.openshift.com/pub/openshift-v4/${ARCH_PPC64LE}/dependencies/rhcos/${OCP_MAJMIN}/${OCP_VERSION}/rhcos-${OCP_VERSION}-${ARCH_PPC64LE}-live.${ARCH_PPC64LE}.iso"
$ export ROOT_FS_URL_PPC64LE="https://mirror.openshift.com/pub/openshift-v4/${ARCH_PPC64LE}/dependencies/rhcos/${OCP_MAJMIN}/${OCP_VERSION}/rhcos-${OCP_VERSION}-${ARCH_PPC64LE}-live-rootfs.${ARCH_PPC64LE}.img"
----

> +

Reviewer: In the following code block and others like it in this file, the code needs to be rewritten to separate out the content that the user will enter in the YAML file and the command to apply the changes. For an example, see steps 1 - 3 in [Creating an InfraEnv resource for hosted control planes on IBM Z](https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/hosted_control_planes/deploying-hosted-control-planes#hcp-ibm-z-infraenv_hcp-deploy-ibmz).

[source,terminal]
----
envsubst <<"EOF" | oc apply -f -
apiVersion: agent-install.openshift.io/v1beta1
kind: AgentServiceConfig
metadata:
 name: agent
spec:
  databaseStorage:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: ${DB_VOLUME_SIZE}
  filesystemStorage:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: ${FS_VOLUME_SIZE}
  osImages:
    - openshiftVersion: "${OCP_VERSION}"
      version: "${OCP_RELEASE_VERSION_X86}"
      url: "${ISO_URL_X86}"
      rootFSUrl: "${ROOT_FS_URL_X8}"
      cpuArchitecture: "${ARCH_X86}"
    - openshiftVersion: "${OCP_VERSION}"
      version: "${OCP_RELEASE_VERSION_PPC64LE}"
      url: "${ISO_URL_PPC64LE}"
      rootFSUrl: "${ROOT_FS_URL_PPC64LE}"
      cpuArchitecture: "${ARCH_PPC64LE}"
EOF
----
+
The above configuration allows you to create ISO for both x86_64 and ppc64le architectures.

> 

Reviewer: The content in lines 68 - 100 should be in its own module.

[id="hcp-configure-dns_{context}"]
== Configure DNS

*.apps.<cluster_name> record can be pointed to either one of the worker nodes where ingress application is hosted, or if you are able to set up a load balancer on top of the worker nodes, point this record to this load balancer. When you are creating a heterogeneous nodepool, please make sure the workers are reachable from each other or keep them in the same network.

.Procedure

> . Create a Hosted Cluster

Reviewer 2: ```suggestion
. Create a hosted cluster 
```

- This instruction is incomplete. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#procedure-step-that-introduces-a-command to know more about writing steps. 
- Normal term is written as "hosted cluster" and its API object is written as "HostedCluster".

+

> Use a multi-architecture release image when creating the cluster to enable support for heterogeneous nodepools. Find the latest multi arch images from here.

Reviewer 2: ```suggestion
Use a multi-architecture release image when creating the cluster to enable support for heterogeneous node pools. Find the latest multi-architecture images from here.
```

- "....from here" is a little vague. You can say something like, by using the following command/example/configuration, and so on.

+
[source,terminal]
----
$ export CLUSTERS_NAMESPACE=<hosted_cluster_namespace>
$ export HOSTED_CLUSTER_NAME="<hosted_cluster_name>”
$ export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}"
$ export BASEDOMAIN=<basedomain>
$ export PULL_SECRET_FILE=$PWD/pull-secret
$ export OCP_RELEASE=<release_image_version>
$ export MACHINE_CIDR=192.168.122.0/24
# Typically the namespace is created by the hypershift-operator 
# but agent cluster creation generates a capi-provider role that
# needs the namespace to already exist
$ oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}

$ hypershift create cluster agent \
    --name=${HOSTED_CLUSTER_NAME} \
    --pull-secret=${PULL_SECRET_FILE} \
    --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \
    --base-domain=${BASEDOMAIN} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN} \
    --release-image=quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE}
----

> 

Reviewer: The content from lines 103 - 147 should be in its own module.

[id="hcp-create-heterogeneous-nodepools_{context}"]
== Create Heterogeneous NodePools

By default, the previous section creates a nodepool on x86_64 architecture, to create heterogeneous nodepools, you can apply the below manifest with the architecture you want.

.Procedure

> . to create heterogeneous nodepools

Reviewer: Be sure to use complete sentences and capitalize the first word in a sentence.
```suggestion
. To create heterogeneous node pools, create a YAML file that contains the following content:
```

+
[source,terminal]
----
apiVersion: hypershift.openshift.io/v1beta1
kind: NodePool
metadata:
  name: ${HOSTED_CLUSTER_NAME}
  namespace: ${CLUSTERS_NAMESPACE}
spec:
  arch: ${ARCH_PPC64LE}
  clusterName: ${HOSTED_CLUSTER_NAME}
  management:
    autoRepair: false
    upgradeType: InPlace
  nodeDrainTimeout: 0s
  nodeVolumeDetachTimeout: 0s
  platform:
    agent:
      agentLabelSelector:
        matchLabels:
          inventory.agent-install.openshift.io/cpu-architecture: ${ARCH_PPC64LE}
    type: Agent
  release:
    image: quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE}
  replicas: 0
----
+
This will create a nodepool of architecture ppc64le with zero replicas.
+
[source,terminal]
----
agentLabelSelector:
    matchLabels:
        inventory.agent-install.openshift.io/cpu-architecture: ppc64le
----
+
This selector block is used to select the agents which are matching the given label. Which will ensure it will select only the agents from ppc64le architecture when it is scaled.

> 

Reviewer: The content from lines 150 - 196 should be in its own module.

[id="hcp-create-infraenv_{context}"]
== Create Infraenv

For heterogeneous nodepools, you need to create infraenv for each architecture you are going to have in your HCP. i.e. for heterogeneous nodepools with x86_64 and ppc64le architecture, you will need to create two infraenvs with both architectures.

.Procedure

> . Create Infraenv

Reviewer 2: ```suggestion
. Create Infraenv
```

This instruction is incomplete

+
[source,terminal]
----
export SSH_PUB_KEY=$(cat $HOME/.ssh/id_rsa.pub)
----
+
[source,terminal]
----
envsubst <<"EOF" | oc apply -f -
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: ${HOSTED_CLUSTER_NAME}-${ARCH_X86}

>   namespace: ${HOSTED_CONTROL_PLANE_NAMESPACE}

Reviewer 2: ```suggestion
  namespace: <hosted_cluster_name>-<arch_x86>
```
You can convert all of the environment variables into user-replaced values (provided there is no mention of `export` commands for env vars). i.e user will replace variables with their custom values. See this guideline to know how to use/write them: https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values

spec:
  cpuArchitecture: ${ARCH_X86}
  pullSecretRef:
    name: pull-secret
  sshAuthorizedKey: ${SSH_PUB_KEY}
EOF
----
+
[source,terminal]
----
envsubst <<"EOF" | oc apply -f -
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: ${HOSTED_CLUSTER_NAME}-${ARCH_PPC64LE}
  namespace: ${HOSTED_CONTROL_PLANE_NAMESPACE}
spec:
  cpuArchitecture: ${ARCH_PPC64LE}
  pullSecretRef:
    name: pull-secret
  sshAuthorizedKey: ${SSH_PUB_KEY}
EOF
----
+
This will create two infraenvs with x86_64 & ppc64le architectures. Before proceeding, make sure that the OS images for both architectures are added to the AgentServiceConfig. After this you can use the above infraenvs to get the minimal iso and follow the create-agent-cluster.md to add them to your cluster as agents.

> 

Reviewer: The content from lines 198 - 201 should be in its own module.

// Module included in the following assemblies:
//
// * hosted_control_planes/hcp-deploy/hcp-deploy-ibm-power.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-ibm-power-create-heterogeneous-nodepools-agent-hc_{context}"]

> = Creating Heterogeneous node pools on Agent hosted clusters

Reviewer 3: ```suggestion
= Creating heterogeneous node pools on Agent hosted clusters
```

> include::hcp-ibm-power-create-heterogeneous-nodepools-agent-hc-con.adoc[leveloffset=+2]

Reviewer: ```suggestion

```
I just realized that you were including a module within another module. I think this might one of the reasons why the tests are failing.

Reviewer 3: Ah. I see why the preview looks odd. See https://github.com/openshift/openshift-docs/pull/92359/files#r2146094888

> You can create heterogeneous node pools on the Agent platform. Below configuration can be used to create ISO for both `x86_64` and `ppc64le` architectures.

Reviewer 3: What is a heterogeneous node pool and why would I create it? Must I create one for both `x86_64` and `ppc64le` architectures? Can I use others?

> You can create heterogeneous node pools on the Agent platform. Below configuration can be used to create ISO for both `x86_64` and `ppc64le` architectures.

Author: We have now provided the explanation for above. Updated as:
```
You can create heterogeneous node pools on the agent platform. It enables clusters to run diverse machine types e.g. `x86_64` or `ppc64le` within a single hosted cluster.
```

.Procedure

> . Configure `AgentServiceConfig` by using two operating system (OS) images with heterogeneous architecture.

Reviewer 3: Do I have to use two operating systems? It is not clear if I have to use  `x86_64` and `ppc64le` architectures, or if those are examples.

spec:
  databaseStorage:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: <db_volume_name> <1>
  filesystemStorage:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: <fs_volume_name> <2>
  osImages:
    - openshiftVersion: <ocp_version> <3>
      version: <ocp_release_version_x86> <4>
      url: <iso_url_x86> <5>
      rootFSUrl: <root_fs_url_x8> <6>
      cpuArchitecture: <arch_x86> <7>
    - openshiftVersion: <ocp_version> <8>
      version: <ocp_release_version_ppc64le> <9>
      url: <iso_url_ppc64le> <10>
      rootFSUrl: <root_fs_url_ppc64le> <11>
      cpuArchitecture: <arch_ppc64le> <12>
EOF
----
+
<1> The Multicluster Engine (MCE) `agentserviceconfig` config, database volume name.
<2> The Multicluster Engine (MCE) `agentserviceconfig` config, filesystem volume name.
<3> The current version of {product-title}.
<4> The current {product-title} release version for x86.
<5> The ISO URL for x86.
<6> The root fs URL for X8.
<7> The CPU architecture for x86.
<8> The current Openshift version.
<9> The Openshift release version for `ppc64le`.
<10> The ISO URL for `ppc64le`. 
<11> The root fs URL for `ppc64le`.

> <12> The CPU architecture for `ppc64le`.

Reviewer 3: Are these values examples or required?

// Module included in the following assemblies:
//
// * hosted_control_planes/hcp-deploy/hcp-deploy-ibm-power.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-ibm-power-create-heterogeneous-nodepools-agent-hc_{context}"]
= Creating heterogeneous node pools on agent hosted clusters

> You can create heterogeneous node pools on the agent platform. It enables clusters to run diverse machine types, for example, `x86_64` or `ppc64le`, within a single hosted cluster. 

Reviewer 3: Do we need to tell users what a heterogeneous node pool is, and why I would use one?

> To create a heterogeneous node pool, you must create an agent service configuration custom resource (CR) that informs the Operator how much storage is needed for components such as the database and filesystem. The CR also defines which {product-title} versions to maintain.

Reviewer 4: Should probably be

```suggestion
To create a heterogeneous node pool, you must create an `AgentServiceConfig` custom resource (CR) that informs the Operator how much storage is needed for components such as the database and filesystem. The CR also defines which {product-title} versions to maintain.
```

> To create a heterogeneous node pool, you must create an agent service configuration custom resource (CR) that informs the Operator how much storage is needed for components such as the database and filesystem. The CR also defines which {product-title} versions to maintain.

Reviewer 3: If it's not too much work, maybe do something like this. 

Make this module a CONCEPT and list the general steps, represented by the other modules. Make the procedure its own module. Does that make sense? 

```
To create a heterogeneous node pool, perform the following general steps, as described in the following sections: 
* Create an agent service configuration custom resource (CR).
* Create an agent cluster.
* Create the heterogeneous node pool.
* Configure DNS for hosted control planes
* Create an `infraenv` custom resource (CR) for each architecture.
* Add agents to the heterogeneous cluster.
```
If any of these steps is optional, you could indicate it in the list to be clear, for example: 

* Optional: Create the heterogeneous node pool.

Author: @mburke5678 I have updated the PR as per above.
```
$ git status
On branch hcp_update
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   modules/hcp-ibm-power-create-heterogeneous-nodepools-agent-hc-con.adoc
        modified:   modules/hcp-ibm-power-create-heterogeneous-nodepools-agent-hc.adoc
```
Request you to take a look. Thanks.

// Module included in the following assemblies:
//
// * hosted_control_planes/hcp-deploy/hcp-deploy-ibm-power.adoc

:_mod-docs-content-type: PROCEDURE
[id="hcp-ibm-power-create-heterogeneous-nodepools-agent-hc_{context}"]
= Creating heterogeneous node pools on agent hosted clusters

You can create heterogeneous node pools on the agent platform. It enables clusters to run diverse machine types e.g. `x86_64` or `ppc64le` within a single hosted cluster. The following example show how to configure a heterogeneous node pool.

.Procedure

. Configure `AgentServiceConfig` with two heterogeneous architecture operating system (OS) images.
+
[Note]
====
Here we must create an agent service configuration custom resource that instructs the operator how much storage is needed for the various components like database and filesystem. It also define the OpenShift versions to maintain.
====
+
[source,terminal]
----

> envsubst <<"EOF" | oc apply -f -

Reviewer 5: ```suggestion
$ envsubst <<"EOF" | oc apply -f -
```
